[WIP!] Update lizard docs (#16)
nens/lizard-docs: source/c_dataexchange.rst

==========================
Data Management & Exchange
==========================

We support multiple types of data uploads.
Data can be uploaded manually, via the data management interface, or you can set up real-time data connections using the API.
We can also provide support on either manual or automatic data uploads. 

.. note::

>     Please note that Lizard assumes the data to be in UTC

Reviewer: ```suggestion
    Please note that Lizard assumes date and time to be in UTC.
```

data/date uit elkaar houden?

Rasters
=======

We offer multiple types of data uploads.
Ranging from file uploads, automatic file uploads to web services.
All options per data type are listed below, including file requirements.  

Requirements 
--------------

Your raster data has to be in the format of a single band, georeferenced TIFF (geotiff), with the following requirements: 

> * **Geotiff should have valid projection** including transformation (EPSG code). All projections supported by proj4 are supported.

Reviewer: ```suggestion
* **Geotiffs should have valid projection** including transformation (EPSG code). All projections supported by proj4 are supported.
```

Idem de regels hieronder.

* **Geotiff should have a NODATA value**.
* **Geotiff should be single band**. RGB or multi-band is not supported. 
* **Temporal raster datasets** with multiple timesteps **should be supplied with a single geotiff per timestamp**

Creating and editing a Raster Store
-------------------------------------

The first step in uploading your raster datasets is to create a Raster Store.
This can easily be done using our Data Management app.
Following this step-by-step tutorial to upload a raster dataset:

.. image:: /images/c_dataexchange_01.jpg

> The Data Management interface is available at: “www.{your_organisation}.lizard.net/management/”.

Reviewer: ```suggestion
The Data Management interface is available at: ``www.{your_organisation}.lizard.net/management/``.
```

After landing on this page, please click on ‘Data Management’, then ‘Rasters’.
Click on “New Raster” |NewRaster| to open the form for new Rasters.
Or choose an existing raster to edit.  

.. |NewRaster| image:: /images/c_dataexchange_02.png

#. Choose the organisation you’re supplying data for. 
#. Choose the organisations you want to share this dataset with. 

> #. Choose the preferred authorization type (read more).

Reviewer: "read more"?

#. Give the dataset a name.
#. Describe your dataset. Make sure to name the source and describe the analysis that resulted in this dataset. Users can read this description in the Lizard Catalog.
#. Choose how your data should be aggregated. This functionality is only needed when you want to use the Region Analysis mode in Lizard Portal or Lizard API. 
#. Choose the observation type of this dataset. 
#. Choose a preferred color map. Choose “Rescalable” if you want to be able to rescale the color map in Lizard Portal.
#. Fill in the supplier name. We use your username by default.
#. You can fill in a supplier code for your own administration.
#. If you’re supplying a temporal dataset. Choose “Raster Series”. Next, fill in the interval of the dataset. 

> #. Click submit. You have now created the Raster Store. You’re all set up to  supply your geotiff’s using the upload button. 

Reviewer: ```suggestion
#. Click submit. You have now created the Raster Store. You’re all set up to  supply your geotiffs using the upload button. 
```

Geen bezits-s.

Upload 
------

> You can supply your GeoTIFF’s in multiple ways: 

Reviewer: ```suggestion
You can supply your GeoTIFFs in multiple ways: 
```

Even zoeken waar die 's nog meer voorkomt.
En even kijken hoe je geotiff/GeoTIFF wil spellen qua hoofdletters.

> * Use the Data Management App 

Reviewer: "Use", maar de koptekst verderop is "Using".

* Use the Lizard API 
* Use the Lizard FTP

Using the Data Management App
++++++++++++++++++++++++++++++

Once you have successfully created a raster store you will see the pop up below.

.. image:: /images/c_dataexchange_03.png

Choose upload data to browse your GeoTIFF’s.
When you want to add data to an existing raster store, click on the upload icon |uploadicon| in the list of existing Raster Stores. 

.. |uploadicon| image:: /images/c_dataexchange_04.png

You can supply multiple rasters, Lizard will blend them together! Click “Save all Files” to start uploading your data.
Your GeoTIFF’s will be uploaded in a task. You can follow the status of the task by clicking “show asynchronous task”.

.. image:: /images/c_dataexchange_05.png

Using the Lizard API
++++++++++++++++++++

.. warning::
   Text required!

Using the Lizard FTP
++++++++++++++++++++ 

.. warning::
   Text required!

> Time Series

Reviewer: ```suggestion
Time series
```

Je hebt het met een spatie geschreven. Even kijken of dat correct is. Als halve Duitser zou ik er zelf "timeseries" van maken.
Als het maar consistent gedaan wordt :-)

    2015-01-01T01:00:00Z, GWmBGS, 248.0, your_own_code_1
    2015-01-01T00:00:00Z, GWmMSL, 252.3, your_own_code_2
    2015-01-02T05:00:00Z, GWmMSL, 234.2, your_own_code_3

The columns should contain:

* **timestamp:** a timestamp in iso8601 format.
* **unit id / name:** This is both the name of the observation type name and this will become the timeseries name.
* **value:** value as either a float or integer number.
* **asset id:** either an asset uuid or an organisation-code (an identifier for an asset, unique for your organisation). In case the assets have been added with a code under category [columns].

Since a csv should not contain a header, your csv should look like this:

.. csv-table:: Example without headers

    2015-01-01T00:00:00Z, GWmMSL, 248.0, your_own_code_1
    2015-01-01T01:00:00Z, GWmBGS, 248.0, your_own_code_1
    2015-01-01T00:00:00Z, GWmMSL, 252.3, your_own_code_2
    2015-01-02T05:00:00Z, GWmMSL, 234.2, your_own_code_3

Authentication
--------------

SFTP users are authenticated with a username / password.

Supported data formats
----------------------

Via SFTP the following formats are supported:

    * CSV

>     * PiXML

Reviewer: Spaties voor de bullets weg.

Every supplier has its own directory on the SFTP.
It can be accessed by logging in with the Lizard credentials.

As soon as a new file is uploaded to the SFTP server, it will be automatically recognised and processed by Lizard.
After processing the file is moved to a backup for a limited period of time.

When a file is rejected, the supplied file is moved to the directory ‘rejected’ and a message is sent to the suppliers Inbox.
In the Inbox a supplier can see the status of his supplied files.

CSV
+++++

Use CSV for supplying timeseries data with numerical or textual values according to the following format:

.. code-block:: none

>     <timestamp>,<timeseries_supplier_code or uuid>,<value>[\n]

Reviewer: Die `[\n]` aan het eind kan beter weg, denk ik.

    <timestamp>,<timeseries_supplier_code or uuid>,<value>[\n]
    <timestamp>,<timeseries_supplier_code or uuid>,<value>[\n]      
    …

Where:

    * **timestamp:** time in UTC in ISO 8601 format. For example 2012-10-26T09:22:35Z. Supplying timestamps in different timezone is only allowed when the UTC offset is added to the timestamp according to ISO 8601. For example: 2012-10-26T07:22:35+02.
    * **timeseries_supplier_code or UUID:** supplier_code attribute of timeseries as registered by administrator/supplier or the UUID of the timeseries object.
    * **value:** numerical or textual value.

>     * **[\\n]:** newline character.

Reviewer: Spaties voor bullets weg.
Geldt ook voor de rest van het bestand. Dit is tot nu toe de enige waarin ik het gezien heb.

    * Use the empirical CSV format where the field separator is a comma (,) and the decimal separator a period (.).

PiXML
+++++++++

Pi-XML is a file exchange format used by Delft-FEWS. A description can be found `here <https://publicwiki.deltares.nl/display/FEWSDOC/Delft-Fews+Published+Interface+timeseries+Format+%28PI%29+Import>`_.

Images
++++++++++++

The following formats are supported for image timeseries:

    * PNG
    * JPEG

Image filenames should have the following format:

.. code-block:: none

    <timeseries_supplier_code>_<timestamp>.png
    <timeseries_supplier_code>_<timestamp>.jpg

Where:

    * **timestamp:** time in UTC in ISO 8601 format. Colons (:) should be excluded as they are not allowed in filenames. An example timestamp is: 2012-10-26T092235Z. Supplying timestamps in different timezone is only allowed when the UTC offset is added to the timestamp according to ISO 8601. For example: 2012-10-26T072235+02.
    * **timeseries_supplier_code:** supplier_code attribute of timeseries as registered by administrator/supplier.

Video and multimedia
++++++++++++++++++++++

The following multimedia file extensions are supported for multimedia timeseries:

>     * PDF

Reviewer: Moet het in hoofdletters zijn? `.pdf` (dus als code, met dubbele backticks eromheen) ziet er denk ik vriendelijker uit.

    * AVI
    * WMV

Multimedia filenames should have the following format:

.. code-block:: none

    <timeseries_supplier_code>_<timestamp>.pdf
    <timeseries_supplier_code>_<timestamp>.avi
    <timeseries_supplier_code>_<timestamp>.wmv

Where:

* **timestamp:** time in UTC in ISO 8601 format or as a unix timestamp in ms since epoch. Colons (:) should be excluded as they are not allowed in filenames. An example timestamp is: 2012-10-26T092235Z. Supplying timestamps in different timezone is only allowed when the UTC offset is added to the timestamp according to ISO 8601. For example: 2012-10-26T072235+02.
* **timeseries_supplier_code:** supplier_code attribute of timeseries as registered by administrator/supplier.

LMW data
+++++++++++

> Every 10 minutes a file is downloaded from LMW.

Reviewer: Afko? Zeker buitenlanders zullen geen flauw idee hebben wat een Local Management Waterboard nou weer is.

Error handling
++++++++++++++++++

> When a file is in the **wrong format**, **authorisation fails** and / or **value type is not valid**:

Reviewer: "in the"? In de wattus?
Even herschrijven.

    * File is moved to ‘rejected’ directory of supplier
    * An error is logged
    * A message is sent to the Inbox of the supplier

SUF-HYD
+++++++++

SUF-HYD files can be imported manually, by uploading a file to https://demo.lizard.net/import/

We currently do not support GWSW-Hyd yet.

The description of SUF-HYD files can be found here: https://www.riool.net/documents/20182/557556/suf-hyd-gegevens%20stelsel-1996-2003.pdf/512c2708-0594-4227-998e-f9c51bec6a50 

API
------------

Timeseries data can be supplied with a POST request to the timeseries data endpoint in the API (`<baseurl>`/api/v3/timeseries/{uuid}/data/).
Interaction with the API can be done from e.g. Postman or Python.

> User credentials should be included in the header and the data in the payload of the request. 

Reviewer: Standaard "Basic auth" headers? Of de custom lizard headers?

Value based timeseries
+++++++++++++++++++++++++++

This type of timeseries consists of integers, floats, float arrays or text. The body of the request is a JSON object with timestamps and values:

.. code-block:: json 

    {
    	"data": [{
    			"datetime": "2019-07-01T01:30:00Z",
    			"value": 40.7
    		},
    		{
    			"datetime": "2019-07-01T02:00:00Z",
    			"value": 39.1
    		}
    	]
    }

File based timeseries
++++++++++++++++++++++

> This type consists of images, movies or files. A single files is posted on a certain datetime, which is included in the header of the request.

Reviewer: ```suggestion
This type consists of images, movies or files. A single file is posted on a certain datetime, which is included in the header of the request.
```

> An example of an upload of an image using requests in Python:

Reviewer: ```suggestion
An example of an upload of an image using ``requests`` in Python:
```

.. code-block:: none  

    import requests
    import datetime as dt

    now = dt.datetime.utcnow()
    uuid = ‘385c08c5-a0cf-4097-a98f-b6f053ef32c6’
    url = 'https://demo.lizard.net/api/v3/timeseries/{}/data/'.format(uuid)
    data = open('./x.png', 'rb').read()
    res = requests.post(url=url,
                        data=data,
                        headers={
                        'Content-Type': 'image/png',
                        'datetime': now.strftime('%Y-%m-%dT%H:%M:%S.%fZ'),

>                         'username': 'jane.doe',

Reviewer: Ah, custom lizard. Dan elders zeggen of basic auth ook werkt.

                        'password': 'janespassword'
                        })

Vectors
=======

We support vector synchronisation.
This type of data feed has to be configured per customer.
Changes in location names, coördinates and new locations can be seen in Lizard as soon as the following day. 

Upload vectors as a shapefile
-----------------------------

Assets can be uploaded to Lizard with shapefiles via the import form at <base-url>/import.
These shapefiles contain information about assets or assets together with their nested assets (e.g. GroundwaterStations and their Filters).

A shapefile can be uploaded as a zipped archive.

> The zipfile should contain at least a .dbf, .shp, .sh xand .ini file.

Reviewer: Dubbele backticks om de formaten.

.. code-block:: python

	import requests
	from localsecret import username, password
	headers = {"username": username, 
			   "password": password}

	def post_geotiff_to_lizard(endpoint, file_path, timestamp=None):
		"""Function to post a temporal geotiff to lizard"""
		url = "{}data/".format(endpoint)
		file = {"file": open(file_path, "rb")}
		if timestamp:
			data = {"timestamp": timestamp}
			r = requests.post(url=url, data=data, files=file, headers=headers)
		else: # use to send data to non-temporal endpoints
			r = requests.post(url=url, files=file, headers=headers)
		return r

	uuid = "b73189fc-058d-4351-9b20-2538248fae4f"
	endpoint = "https://demo.lizard.net/api/v4/rasters/{}/".format(uuid)
	file_path = "local_geotiff.tif"
	timestamp = "2020-01-01T000000Z"

	response = post_temporal_geotiff_to_lizard(endpoint, file_path, timestamp)

Using the Lizard FTP
++++++++++++++++++++ 

The examples below elaborate on how to supply raster data to the Lizard FTP and how to manage your data flow to Lizard.
Because data uploads depend on the system configuration of the data provider, we provide a code example to make a tool for automatic uploads.

> * You can do this using the python FTPlib package (example 1).

Reviewer: Effe entertje ervoor.

You can copy paste this code in your own .ini file and zip it together with the shapefile.

SUF-HYD
+++++++++

SUF-HYD files can be imported manually, by uploading a file to https://demo.lizard.net/import/

We currently do not support GWSW-Hyd yet.

The description of SUF-HYD files can be found here: https://www.riool.net/documents/20182/557556/suf-hyd-gegevens%20stelsel-1996-2003.pdf/512c2708-0594-4227-998e-f9c51bec6a50 

Data downloads
==============

Rasters
-------

Download of rasters is possible but limited via the Lizard portal.

> The current limit is a 1 milion by 1 milion (10^9) pixels download.

Reviewer: milion => miLLion

En da's 10^6 (of 10^12 als je het totale aantal pixels bedoeld).
10^6 is niet nodig voor miljoen, want da's altijd duidelijk. (wel bij billion of miliard, helaas)

Only possible when you are zoomed in far enough, depending on the resolution of the specific raster.

> Select a raster from the datalayers menu to the right.

Reviewer: Lijstje van maken?

Author: Dit is voor nu beschreven voor de Lizard portal zelf. Binnenkort (nu al volgens mij) is raster export via de Lizard Catalogue de beste manier om dit te gebruiken. Alleen moet die tekst nog geschreven worden. Dat zal komende week wel gebeuren, dus dan updaten we dit sowieso.

Zoom in to the required extent.
Click the export button, and click on the Rasters tab in the Export Data window.
Select the required projection and cel size.
Click on Start Export.
When raster export is done, a download link will be supplied via the Lizard inbox.

Timeseries
-----------

> Lizard supports two types of timeseries.

Reviewer: Lijstje van maken?

Author: Misschien later.
